osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/infra/networking/Mx382AlarmDecoder.java

```diff
@@ -0,0 +1,125 @@
+  }
+
+  private void skipDateTime(final InputStream inputStream) throws UnrecognizedMessageDataException {
+    // The first byte of the datetime is the length byte
+    final byte dateTimeLength = this.readByte(inputStream);
+    this.builder.appendByte(dateTimeLength);
+    // Skip the rest of the datetime bytes
+    final byte[] dateTimeBytes = this.readBytes(inputStream, dateTimeLength);
+    this.builder.appendBytes(dateTimeBytes);
+  }
+
+  private void decodeEquipmentIdentifier(final InputStream inputStream)
+      throws UnrecognizedMessageDataException {
+
+    // Check bytes of incoming mx382 message
+    this.checkBytes(inputStream, DEVICE_ID_2_CLASS_ID, "class-id");
+    this.checkBytes(inputStream, AMM_FORWARDED_ALARM_VERSION_0, "logical name forwarded alarm");
+    this.checkByte(inputStream, VALUE_ATTRIBUTE_ID, "attribute-id");
+
+    // Read the equipment identifier from incoming mx382 message
+    final byte[] equipmentIdentifierBytes =
+        this.readBytes(inputStream, DSMR22_EQUIPMENT_IDENTIFIER_LENGTH);
+    this.builder.withEquipmentIdentifier(
+        new String(equipmentIdentifierBytes, StandardCharsets.US_ASCII));
+  }
+
+  private void addAlarm() {
+    this.builder.withTriggerType(PUSH_ALARM_TRIGGER);
+    final Set<AlarmTypeDto> alarmSet = new HashSet<>();
+    alarmSet.add(AlarmTypeDto.LAST_GASP);
+    this.builder.addAlarms(alarmSet);
+  }
+
+  private void checkByte(final InputStream inputStream, final byte expectedByte, final String name)
+      throws UnrecognizedMessageDataException {
+    final byte readByte = this.readByte(inputStream);
+    this.builder.appendByte(readByte);
+    if (readByte != expectedByte) {
+      throw new UnrecognizedMessageDataException(
+          String.format(
+              "Expected a %s (%s), but encountered: (%s)",
+              name, toHexString(expectedByte), toHexString(readByte)));
+    }
+  }
+
+  private void checkBytes(
+      final InputStream inputStream, final byte[] expectedBytes, final String name)
+      throws UnrecognizedMessageDataException {
+    final byte[] readBytes = this.readBytes(inputStream, expectedBytes.length);
+    this.builder.appendBytes(readBytes);
+    if (!Arrays.equals(readBytes, expectedBytes)) {
+      throw new UnrecognizedMessageDataException(
+          String.format(
+              "Expected a %s (%s), but encountered: (%s)",
+              name, toHexString(expectedBytes), toHexString(readBytes)));
+    }
+  }
+
+  private static String toHexString(final byte b) {
+    return toHexString(new byte[] {b});
+  }
+
+  public static String toHexString(final byte[] ba) {
+    final StringBuilder sb = new StringBuilder("[");
+    boolean first = true;
+    for (final byte b : ba) {
+      if (!first) {
+        sb.append(", ");
+      } else {
+        first = false;
+      }
+      sb.append("0x");
+      sb.append(Hex.toHexString(new byte[] {b}));
+    }
+    sb.append("]");
+    return sb.toString();
+  }
+}
```

Comment:
Why not use `Hex.toHexString(ba)`?

Reply:
Hex.toHexString(ba) will result in 0180. Desired formatting is [0x00, 0x80]